File: backend/app/services/scanner.py
from app.core.store import new_id, now, store
from app.schemas import ScanLog, ScannerValidate, ScanState, TicketStatus
from app.services.audit import log_audit
from app.services.qr import hash_qr_token, verify_qr_token


def validate_scan(payload: ScannerValidate) -> dict:
    valid, qr_payload, reason = verify_qr_token(payload.qr_token)
    state = ScanState.INVALID
    message = reason or "Token not found"
    ticket = None
    if valid and qr_payload:
        token_hash = hash_qr_token(payload.qr_token)
        token = next((item for item in store.ticket_tokens if item.token_hash == token_hash), None)
        ticket = next((item for item in store.tickets if token and item.id == token.ticket_id), None)
        if not ticket:
            state = ScanState.INVALID
            message = "Ticket record missing"
        elif qr_payload.get("ticket_id") != ticket.id or qr_payload.get("event_id") != ticket.event_id:
            state = ScanState.INVALID
            message = "QR payload does not match ticket"
        elif payload.event_id != "auto" and ticket.event_id != payload.event_id:
            state = ScanState.WRONG_EVENT
            message = "Ticket belongs to another event"
        elif ticket.status == TicketStatus.CANCELLED:
            state = ScanState.CANCELLED
            message = "Ticket is cancelled"
        elif ticket.status == TicketStatus.REFUNDED:
            state = ScanState.REFUNDED
            message = "Ticket is refunded"
        elif ticket.status == TicketStatus.EXPIRED:
            state = ScanState.EXPIRED
            message = "Ticket is expired"
        elif ticket.status == TicketStatus.CHECKED_IN:
            state = ScanState.ALREADY_USED
            message = f"Already checked in at {ticket.checked_in_at}"
        else:
            state = ScanState.VALID
            message = "Entry allowed"
            ticket.status = TicketStatus.CHECKED_IN
            ticket.checked_in_at = now()
            ticket.checked_in_gate_id = payload.gate_id
            ticket.checked_in_by = payload.scanner_user_id
    log = ScanLog(
        id=new_id("scan"),
        event_id=ticket.event_id if ticket else payload.event_id,
        ticket_id=ticket.id if ticket else None,
        scanner_user_id=payload.scanner_user_id,
        gate_id=payload.gate_id,
        device_id=payload.device_id,
        scan_result=state,
        metadata={"message": message},
        scan_time=now(),
    )
    store.scan_logs.insert(0, log)
    log_audit("ticket.scanned", "scan", log.id, event_id=log.event_id, metadata={"result": state, "ticket_id": log.ticket_id})
    if state == ScanState.ALREADY_USED:
        log_audit("scan.duplicate_attempted", "ticket", ticket.id if ticket else "unknown", event_id=log.event_id)
    return {"status": state, "message": message, "ticket_id": log.ticket_id, "scanned_at": log.scan_time}


def manual_lookup(query: str) -> list:
    q = query.lower()
    return [
        ticket for ticket in store.tickets
        if q in ticket.id.lower()
        or q in ticket.attendee_name.lower()
        or q in ticket.attendee_phone.lower()
        or (ticket.attendee_email and q in ticket.attendee_email.lower())
    ]

File: backend/modules/scanner.ts
import { createId, nowIso } from "../core/ids";
import { getStore, persistStoreRecord, persistStoreUpdate } from "../core/store";
import { getTicket } from "./tickets";
import { recordAudit } from "./audit";
import { verifyQrToken } from "./qr-service";

function normalizeTicketStatus(status: string | undefined) {
  if (!status) return "active";
  if (status === "issued") return "active";
  if (status === "checked_in") return "used";
  return status;
}

export async function validateScan(payload: Record<string, unknown>) {
  const store = getStore();
  const token = String(payload.qrToken ?? payload.ticketId ?? "");
  const verified = payload.qrToken ? verifyQrToken(token) : { valid: Boolean(token), ticketId: token, eventId: "" };

  // ── Ticket resolution (3-step, most-specific first) ──────────────────────
  // Step 1: direct qrToken field match (exact signed payload stored at issue time)
  const byQrToken = store.tickets.find((t) => t.qrToken === token);
  // Step 2: verified payload ticket_id (from decoded HMAC token)
  const byVerifiedId = verified.ticketId ? getTicket(verified.ticketId) : undefined;
  // Step 3: raw input treated as ticket ID (manual entry fallback)
  const byRawId = getTicket(token);
  const ticket = byQrToken ?? byVerifiedId ?? byRawId;
  // Re-evaluate validity: if we found by qrToken match, token is implicitly valid
  const isValid = verified.valid || Boolean(byQrToken);

  const scanTime = nowIso();
  const logScan = async (scanResult: string, message: string) => {
    await persistStoreRecord("scanLogs", {
      id: createId("scan"),
      organizationId: ticket?.organizationId,
      eventId: ticket?.eventId,
      ticketId: ticket?.id ?? verified.ticketId ?? token,
      qrToken: token,
      scannerUserId: String(payload.scannerUserId ?? "usr_scanner_demo"),
      gateId: String(payload.gateId ?? "gate_main"),
      deviceId: String(payload.deviceId ?? "web"),
      scanResult,
      message,
      scanTime,
    });
  };

  if (!ticket || !isValid) {
    await logScan("INVALID", "Ticket not found");
    return { status: "INVALID", message: "Ticket not found", ticketId: token.length < 30 ? token : undefined };
  }

  const attendeeName = String(ticket.attendeeName ?? ticket.userId ?? "—");

  const currentStatus = normalizeTicketStatus(String(ticket.status));
  if (["cancelled", "refunded", "expired"].includes(currentStatus)) {
    const result = currentStatus.toUpperCase();
    await logScan(result, `Ticket is ${currentStatus}`);
    recordAudit({ action: "wallet.qr.scan.rejected", entityType: "tickets", entityId: ticket.id, newValue: { result } });
    return { status: result, ticketId: ticket.id, attendeeName, message: `Ticket is ${currentStatus}` };
  }

  if (currentStatus === "used") {
    await logScan("ALREADY USED", "Ticket already used");
    recordAudit({ action: "wallet.qr.scan.duplicate", entityType: "tickets", entityId: ticket.id });
    return {
      status: "ALREADY USED",
      ticketId: ticket.id,
      attendeeName,
      checkedInAt: ticket.scannedAt ?? ticket.checkedInAt,
      gateName: "Main Gate",
      scannerStaffName: ticket.scannedBy ?? ticket.checkedInBy,
      deviceId: payload.deviceId,
      message: "Ticket already used",
    };
  }

  ticket.status = "used";
  ticket.scannedAt = scanTime;
  ticket.scannedBy = String(payload.scannerUserId ?? "usr_scanner_demo");
  ticket.checkedInAt = scanTime;
  ticket.checkedInGateId = String(payload.gateId ?? "gate_main");
  ticket.checkedInBy = String(payload.scannerUserId ?? "usr_scanner_demo");

  await persistStoreUpdate("tickets", ticket);
  store.scanLogs.unshift({
    id: createId("scan"),
    organizationId: ticket.organizationId,
    eventId: ticket.eventId,
    ticketId: ticket.id,
    scannerUserId: String(payload.scannerUserId ?? "usr_scanner_demo"),
    gateId: String(payload.gateId ?? "gate_main"),
    deviceId: String(payload.deviceId ?? "web"),
    scanResult: "VALID",
    scanTime,
  });
  await logScan("VALID", "Entry allowed");
  recordAudit({ action: "wallet.qr.scanned", entityType: "tickets", entityId: ticket.id, newValue: { result: "VALID", deviceId: payload.deviceId } });
  return { status: "VALID", ticketId: ticket.id, attendeeName, checkedInAt: scanTime, gateName: "Main Gate", message: "Entry allowed" };
}
